## NumPy

[NumPy](https://numpy.org/doc/stable/index.html) 是一个用于科学计算的 Python 库，提供多维数组对象、各种派生对象（例如掩码数组和矩阵）以及用于快速操作数组的各种例程，包括数学、逻辑、形状操作、排序、选择、I/O、离散傅里叶变换、基本线性代数、基本统计运算、随机模拟等等。

In [3]:
# !pip install numpy
import numpy as np
print(np.__version__)

2.3.2


0.5772156649015329

### 常量

|Constant        |Desc                                                        |
|----------------|------------------------------------------------------------|
|`np.pi`         |圆周率 $\pi=3.141592653589793\cdots$|
|`np.e`          |欧拉常数，自然对数，纳皮尔常数 $e=2.718281828459045\cdots$|
|`np.euler_gamma`|欧拉-马斯刻若尼常数 $\gamma=0.5772156649015329\cdots$|
|`np.inf`        |IEEE 754 正无穷大的浮点表示 $\infty$|
|`np.nan`        |IEEE 754 非数的浮点表示，Not a Number|
|`np.newaxis`    |`None` 的别名，常用于索引数组|

-----

### ndarray

> https://numpy.org/doc/2.3/reference/arrays.html

**ndarray** 是同构数据类型的多维数组，数组的维度也称为轴（axis）。

|Attribute          |Desc                                                      |
|-------------------|----------------------------------------------------|
|`ndarray.ndim`     |数组的轴(axis)数|
|`ndarray.shape`    |数组的形状，一个整数元组，表示数组在每个维度上的大小|
|`ndarray.size`     |数组元素的总数，等于形状元素的乘积|
|`ndarray.dtype`    |数组元素的类型|
|`ndarray.itemsize` |数组中每个元素的大小（以字节为单位），等同于`ndarray.dtype.itemsize`|
|`ndarray.data`     |包含数组实际元素的缓冲区|

-----

### 创建数组（ndarray creation）

> https://numpy.org/doc/2.3/reference/routines.array-creation.html

#### From existing data

|API                                            |Desc                                                           |
|-----------------------------------------------|---------------------------------------------------------------|
|`np.array(array_like[,dtype])`                 |创建一个数组|

#### From shape or value

|API                                            |Desc                                                           |
|-----------------------------------------------|---------------------------------------------------------------|
|`np.empty(shape[,dtype])`                      |创建一个未初始化数组|
|`np.ones(shape[,dtype])`                       |创建一个全 $1$ 数组|
|`np.zeros(shape[,dtype])`                      |创建一个全 $0$ 数组|
|`np.full(shape, fill_value[,dtype])`           |创建一个全 `fill_value` 数组|
|`np.identity(n[,dtype])`                       |创建一个恒等方阵 $\boldsymbol{I}$|
|`np.diag(array_like[,k])`                      |创建一个对角方阵，方阵的对角线元素由 `array_like` 提供|
|`np.eye(N[,M,k,dtype])`                        |创建一个 $N \times M$ 的对角矩阵，其中对角元素是 $1$，其余位置都是 $0$|
|`np.tri(N[,M,k,dtype])`                        |创建一个 $N \times M$ 的矩阵，其中对角线及其下方元素为 $1$，其他位置都是 $0$|
|`np.tril(matrix_like[,k])`                     |创建一个下三角矩阵|
|`np.triu(matrix_like[,k])`                     |创建一个上三角矩阵|

#### Numerical ranges (1D)

|API                                            |Desc                                                           |
|-----------------------------------------------|---------------------------------------------------------------|
|`np.arange([start,]stop[,step,][,dtype])`      |创建一个等差序列数组，类似于 Python 内置的 `range()` 函数|
|`np.linspace(start,stop[,num,endpoint,dtype])` |创建一个在 $[\text{start}, \text{end}]$ 之间均匀分布 `num` 个元素的数组|

-----

### 副本(copy)和视图(view)

> https://numpy.org/doc/2.3/user/basics.copies.html#basics-copies-and-views

**ndarray** 由两部分组成：保存实际数据的**连续数据缓冲区**和数据缓冲区相关信息的**元数据**。

- **视图(view)**：适用于只需要修改元数据的场景，对视图的元素修改都会反应到原始数据缓冲区。`ndarray.view()` 创建原数组的视图。

- **副本(copy)**：复制数据缓冲区和元数据创建的新数组，对副本的任何变更都不会反应到原数组上。`ndarray.copy()` 创建原数组的副本。

- 数组的 `base` 属性可以判断数组是视图还是副本，视图的 `base` 属性返回原始数组，副本的 `base` 属性返回 `None`。

-----